File: kancore_cpp/animeEffect/animationgraphic.cpp
#include "animationgraphic.h"

/**
 * @brief 动画构造函数
 *
 * @param parent        父
 * @param inputItem     加特效的对象
 * @param animePosOn    位移特效是否开(默认开)
 * @param animeScaleOn  缩放特效是否开(默认开)
 * @param animeRotateOn 旋转特效是否开(默认开)
 * @param animeOpacOn   透明度特效是否开(默认开)
 */
AnimationGraphic::AnimationGraphic(QObject *parent, PicGraphic *inputItem, bool animePosOn,
                                   bool animeScaleOn, bool animeRotateOn, bool animeOpacOn) :
    QObject(parent)
{
    item=inputItem;
    /*动画初始化*/
    animePosIsOn   =false;
    animeScaleIsOn =false;
    animeRotateIsOn=false;
    animeOpacIsOn  =false;
    setAnimation_OnOff(animePosOn,animeScaleOn,animeRotateOn,animeOpacOn);
}

AnimationGraphic::~AnimationGraphic()
{
    if(animePosIsOn)
        delete animePos;
    if(animeScaleIsOn)
        delete animeScale;
    if(animeRotateIsOn)
        delete animeRotate;
    if(animeOpacIsOn)
        delete animeOpacity;
}


/**
 * @brief 各种动画开启或关闭,可以重复调用,已集成内存管理,不会内存泄露
 *
 * @param animePosOn    位移动画是否使能
 * @param animeScaleOn  大小动画是否使能
 * @param animeRotateOn 旋转动画是否使能
 * @param animeOpacOn   透明动画是否使能
 */
void AnimationGraphic::setAnimation_OnOff(bool animePosOn, bool animeScaleOn,
                                          bool animeRotateOn, bool animeOpacOn)
{
    /*位移动画开关*/
    if(!animePosIsOn && animePosOn)
    {
        animePos=new QPropertyAnimation(item,"pos");
        connect(animePos,SIGNAL(finished()),this,SLOT(animeFinishedPos()));
        animePosIsOn=true;
    }else if(animePosIsOn && !animePosOn)
    {
        delete animePos;
        disconnect(animePos,SIGNAL(finished()),this,SLOT(animeFinishedPos()));
        animePosIsOn=false;
    }
    /*缩放动画开关*/
    if(!animeScaleIsOn && animeScaleOn)
    {
        animeScale=new QPropertyAnimation(item,"scale");
        connect(animeScale,SIGNAL(finished()),this,SLOT(animeFinishedScale()));
        animeScaleIsOn=true;
    }else if(animeScaleIsOn && !animeScaleOn)
    {
        delete animeScale;
        disconnect(animeScale,SIGNAL(finished()),this,SLOT(animeFinishedScale()));
        animeScaleIsOn=false;
    }
    /*旋转动画开关*/
    if(!animeRotateIsOn && animeRotateOn)
    {
        animeRotate=new QPropertyAnimation(item,"rotation");
        connect(animeRotate,SIGNAL(finished()),this,SLOT(animeFinishedRotate()));
        animeRotateIsOn=true;
    }else if(animeRotateIsOn && !animeRotateOn)
    {
        delete animeRotate;
        disconnect(animeRotate,SIGNAL(finished()),this,SLOT(animeFinishedRotate()));
        animeRotateIsOn=false;
    }
    /*透明度动画开关*/
    if(!animeOpacIsOn && animeOpacOn)
    {
        animeOpacity=new QPropertyAnimation(item,"opacity");
        connect(animeOpacity,SIGNAL(finished()),this,SLOT(animeFinishedOpac()));
        animeOpacIsOn=true;
    }else if(animeOpacIsOn && !animeOpacOn)
    {
        delete animeOpacity;
        disconnect(animeOpacity,SIGNAL(finished()),this,SLOT(animeFinishedOpac()));
        animeOpacIsOn=false;
    }

}






/****************************Public******************************/

/**
 * @brief 位移动画
 *
 * @param startPos    起始点
 * @param endPoint    结束点
 * @param msec        时长
 * @param aTyte       曲线
 * @param animeHandle 句柄
 * @param loopTime    循环次数(-1为无限循环,区别于timeline的0为无限循环)
 */
void AnimationGraphic::gradualPos(QPointF startPos, QPointF endPos, int msec,
                                  const QEasingCurve &aTyte, int animeHandle, int loopTime)
{
    handlerCurPos=animeHandle;
    animePos->setEasingCurve(aTyte);
    animePos->setDuration(msec);
    animePos->setStartValue(startPos);
    animePos->setEndValue(endPos);
    animePos->setLoopCount(loopTime);
    animePos->start();
    emit signAnimeStartPos(animeHandle);
}

/**
 * @brief 大小动画
 *
 * @param startScale 起始比例
 * @param endScale   结束比例
 * @param msec       时长
 * @param aTyte      曲线
 * @param animeHandle句柄
 * @param loopTime    循环次数(-1为无限循环,区别于timeline的0为无限循环)
 */
void AnimationGraphic::gradualScale(qreal startScale,qreal endScale,int msec,
                                    const QEasingCurve &aTyte,int animeHandle,int loopTime)
{
    handlerCurScale=animeHandle;
    animeScale->setEasingCurve(aTyte);
    animeScale->setDuration(msec);
    animeScale->setStartValue(startScale);
    animeScale->setEndValue(endScale);
    animeScale->setLoopCount(loopTime);
    animeScale->start();
    emit signAnimeStartScale(animeHandle);
}

/**
 * @brief 旋转动画
 *
 * @param startAngle  起始角度
 * @param endAngle    结束角度
 * @param msec        时长
 * @param aTyte       曲线
 * @param animeHandle 句柄
 * @param loopTime    循环次数(-1为无限循环,区别于timeline的0为无限循环)
 */
void AnimationGraphic::gradualRotate(qreal startAngle, qreal endAngle, int msec,
                                     const QEasingCurve &aTyte, int animeHandle,int loopTime)
{
    handlerCurRotate=animeHandle;
    animeRotate->setEasingCurve(aTyte);
    animeRotate->setDuration(msec);
    animeRotate->setStartValue(startAngle);
    animeRotate->setEndValue(endAngle);
    animeRotate->setLoopCount(loopTime);
    animeRotate->start();
    emit signAnimeStartRotate(animeHandle);
}

/**
 * @brief 透明度动画
 *
 * @param startOpac  起始透明度
 * @param endOpac    结束透明度
 * @param msec       时长
 * @param aTyte      曲线
 * @param animeHandle句柄
 * @param loopTime    循环次数(-1为无限循环,区别于timeline的0为无限循环)
 */
void AnimationGraphic::gradualOpac(qreal startOpac, qreal endOpac, int msec,
                                   const QEasingCurve &aTyte, int animeHandle,int loopTime)
{
    handlerCurOpac=animeHandle;
    animeOpacity->setEasingCurve(aTyte);
    animeOpacity->setDuration(msec);
    animeOpacity->setStartValue(startOpac);
    animeOpacity->setEndValue(endOpac);
    animeOpacity->setLoopCount(loopTime);
    animeOpacity->start();
    emit signAnimeStartOpac(animeHandle);
}



/*******************************************slots**************************************************/
/**
 * @brief 动画结束slot
 *
 */
void AnimationGraphic::animeFinishedPos()
{
    emit signAnimeFinishPos(handlerCurPos);
}
void AnimationGraphic::animeFinishedScale()
{
    emit signAnimeFinishScale(handlerCurScale);
}
void AnimationGraphic::animeFinishedRotate()
{
    emit signAnimeFinishRotate(handlerCurRotate);
}
void AnimationGraphic::animeFinishedOpac()
{
    emit signAnimeFinishOpac(handlerCurOpac);
}

File: kancore_py/animeEffect/AnimationGraphic.py
from enum import Enum
from PyQt5.QtCore import QObject, pyqtSignal, pyqtSlot
from PyQt5.QtCore import QPropertyAnimation


class animeType(Enum):
    TypePos = 0
    TypeScale = 1
    TypeRotate = 2
    TypeOpacity = 3


class AnimationGraphic(QObject):
    signAnimeStartPos = pyqtSignal(int)
    signAnimeStartScale = pyqtSignal(int)
    signAnimeStartRotate = pyqtSignal(int)
    signAnimeStartOpac = pyqtSignal(int)
    signAnimeFinishPos = pyqtSignal(int)
    signAnimeFinishScale = pyqtSignal(int)
    signAnimeFinishRotate = pyqtSignal(int)
    signAnimeFinishOpac = pyqtSignal(int)
    animePosIsOn = True
    animeScaleIsOn = True
    animeRotateIsOn = True
    animeOpacIsOn = True
    handlerCurPos = 0
    handlerCurScale = 0
    handlerCurRotate = 0
    handlerCurOpac = 0
    animeOpacity = None
    animePos = None
    animeScale = None
    animeRotate = None
    item = None

    def __init__(self, parent, inputItem, animePosOn, animeScaleOn, animeRotateOn, animeOpacOn):
        super(AnimationGraphic, self).__init__(parent)
        self.item = inputItem
        self.animePosIsOn = False
        self.animeScaleIsOn = False
        self.animeRotateIsOn = False
        self.animeOpacIsOn = False
        self.setAnimation_OnOff(animePosOn, animeScaleOn, animeRotateOn, animeOpacOn)


def setAnimation_OnOff(self, animePosOn, animeScaleOn, animeRotateOn, animeOpacOn):
    if not self.self.animePosIsOn and animePosOn:
        self.animePos = QPropertyAnimation(self.item, "pos")
        self.animePos.finished.connect(self.animeFinishedPos)
        self.animePosIsOn = True
    elif self.animePosIsOn and not animePosOn:
        self.animePos.finished.disconnect(self.animeFinishedPos)
        self.animePosIsOn = False

    if not self.animeScaleIsOn and animeScaleOn:
        self.animeScale = QPropertyAnimation(self.item, "scale")
        self.animeScale.finished.connect(self.animeFinishedScale)
        self.animeScaleIsOn = True
    elif self.animeScaleIsOn and not animeScaleOn:
        self.animeScale.finished.disconnect(self.animeFinishedScale)
        self.animeScaleIsOn = False

    if not self.animeRotateIsOn and animeRotateOn:
        self.animeRotate = QPropertyAnimation(self.item, "rotation")
        self.animeRotate.finished.connect(self.animeFinishedRotate)
        self.animeRotateIsOn = True
    elif self.animeRotateIsOn and not animeRotateOn:
        self.animeRotate.finished.disconnect(self.animeFinishedRotate)
        self.animeRotateIsOn = False

    if not self.animeOpacIsOn and animeOpacOn:
        self.animeOpacity = QPropertyAnimation(self.item, "opacity")
        self.animeOpacity.finished.connect(self.animeFinishedOpac)
        self.animeOpacIsOn = True
    elif self.animeOpacIsOn and not animeOpacOn:
        self.animeOpacity.finished.disconnect(self.animeFinishedOpac)
        self.animeOpacIsOn = False

    def gradualPos(self, startPos, endPos, msec, aTyte, animeHandle, loopTime):
        self.handlerCurPos = animeHandle
        self.animePos.setEasingCurve(aTyte)
        self.animePos.setDuration(msec)
        self.animePos.setStartValue(startPos)
        self.animePos.setEndValue(endPos)
        self.animePos.setLoopCount(loopTime)
        self.animePos.start()
        self.signAnimeStartPos.emit(animeHandle)

    def gradualScale(self, startScale, endScale, msec, aTyte, animeHandle, loopTime):
        self.handlerCurScale = animeHandle
        self.animeScale.setEasingCurve(aTyte)
        self.animeScale.setDuration(msec)
        self.animeScale.setStartValue(startScale)
        self.animeScale.setEndValue(endScale)
        self.animeScale.setLoopCount(loopTime)
        self.animeScale.start()
        self.signAnimeStartScale.emit(animeHandle)

    def gradualRotate(self, startAngle, endAngle, msec, aTyte, animeHandle, loopTime):
        self.handlerCurRotate = animeHandle
        self.animeRotate.setEasingCurve(aTyte)
        self.animeRotate.setDuration(msec)
        self.animeRotate.setStartValue(startAngle)
        self.animeRotate.setEndValue(endAngle)
        self.animeRotate.setLoopCount(loopTime)
        self.animeRotate.start()
        self.signAnimeStartRotate.emit(animeHandle)

    def gradualOpac(self, startOpac, endOpac, msec, aTyte, animeHandle, loopTime):
        self.handlerCurOpac = animeHandle
        self.animeOpacity.setEasingCurve(aTyte)
        self.animeOpacity.setDuration(msec)
        self.animeOpacity.setStartValue(startOpac)
        self.animeOpacity.setEndValue(endOpac)
        self.animeOpacity.setLoopCount(loopTime)
        self.animeOpacity.start()
        self.signAnimeStartOpac.emit(animeHandle)

    @pyqtSlot()
    def animeFinishedPos(self):
        self.signAnimeFinishPos.emit(self.handlerCurPos)

    @pyqtSlot()
    def animeFinishedScale(self):
        self.signAnimeFinishScale.emit(self.handlerCurScale)

    @pyqtSlot()
    def animeFinishedRotate(self):
        self.signAnimeFinishRotate.emit(self.handlerCurRotate)

    @pyqtSlot()
    def animeFinishedOpac(self):
        self.signAnimeFinishOpac.emit(self.handlerCurOpac)
